core/renderers/measurables/rows.js

```diff
@@ -0,0 +1,389 @@
+ * @return {Blockly.blockRendering.InRowSpacer} The last spacer element on
+ *   this row.
+ * @package
+ */
+Blockly.blockRendering.Row.prototype.getLastSpacer = function() {
+  /* eslint-disable-next-line max-len */
+  return /** @type {Blockly.blockRendering.InRowSpacer} */ (this.elements[this.elements.length - 1]);
+};
+
+/**
+ * An object containing information about what elements are in the top row of a
+ * block as well as sizing information for the top row.
+ * Elements in a top row can consist of corners, hats, spacers, and previous
+ * connections.
+ * @param {!Blockly.BlockSvg} block The block for which this represents the top
+ *     row.
+ * @package
+ * @constructor
+ * @extends {Blockly.blockRendering.Row}
+ */
+Blockly.blockRendering.TopRow = function(block) {
+  Blockly.blockRendering.TopRow.superClass_.constructor.call(this);
+
+  this.type = 'top row';
+
+  /**
+   * The starting point for drawing the row, in the y direction.
+   * This allows us to draw hats and simliar shapes that don't start at the
+   * origin. Must be non-negative (see #2820).
+   * @package
+   * @type {number}
+   */
+  this.startY = 0;
+
+  /**
+   * Whether the block has a previous connection.
+   * @package
+   * @type {boolean}
+   */
+  this.hasPreviousConnection = !!block.previousConnection;
+
+  /**
+   * The previous connection on the block, if any.
+   * TODO: Should this be the connection measurable instead? It would add some
+   * indirection but would mean we aren't mixing connections and connection
+   * measurables.
+   * @type {Blockly.RenderedConnection}
+   */
+  this.connection = block.previousConnection;
+
+  var precedesStatement = block.inputList.length &&
+      block.inputList[0].type == Blockly.NEXT_STATEMENT;
+
+  // This is the minimum height for the row. If one of its elements has a
+  // greater height it will be overwritten in the compute pass.
+  if (precedesStatement && !block.isCollapsed()) {
+    this.height = Blockly.blockRendering.constants.LARGE_PADDING;
+  } else {
+    this.height = Blockly.blockRendering.constants.MEDIUM_PADDING;
+  }
+};
+goog.inherits(Blockly.blockRendering.TopRow, Blockly.blockRendering.Row);
+
+/**
+ * Convenience method to get the measurable representing the previous
+ * connection, if one exists.
+ * @return {Blockly.blockRendering.PreviousConnection} The measurable that
+ *     represents the previous connection on this block, or null.
+ * @package
+ */
+Blockly.blockRendering.TopRow.prototype.getPreviousConnection = function() {
+  if (this.hasPreviousConnection) {
+    return /** @type {Blockly.blockRendering.PreviousConnection} */ (this.elements[2]);
+  }
+  return null;
+};
+
+/**
+ * @override
+ */
+Blockly.blockRendering.TopRow.prototype.measure = function() {
+  for (var e = 0; e < this.elements.length; e++) {
+    var elem = this.elements[e];
+    this.width += elem.width;
+    if (!(elem.isSpacer())) {
+      if (elem.type == 'hat') {
```

Comment:
Similar idea for bottom row. You could call super.measure() and then just loop through the elements to get the hat if isHat is true and adjust the startY and height. Since the number of elements in this row is expected to be very small, looping through the elements a second time shouldn't be a performance concern.
An idea to go with this is that topRow could have an additional field that is a reference to the hat element so you don't need to loop though.

A related question, why is startY for the row also set in createTopRow? should that be removed?

Reply:
`startY` for the row isn't set in `createTopRow` (now `populateTopRow`).  That sets `startY` on the block instead.

The row doesn't know about the hat until this measure pass, because the hat is added by pushing directly to the elements array.  But I could definitely change that.

I don't think that I can just figure out the correct height in a later loop, because if the hat was the tallest thing in the block it would override the height of anything else (with the Math.max call) and we wouldn't know what the second-tallest thing was to reset the height to.